### PR TITLE
#128 add row select mechanism

### PR DESCRIPTION
**Pull request by joshxg**
#128 partly - enough for now

- I went with a quick implementation with the `useImperativeHandle` hook to see if we like it. It's against the 'react grain' (passing props down, rather than passing a thing down to access the child) so I can see it not being liked. I am still undecided
- couple offroads
- Think best to change the icons of the checkbox for the nice border radius (which I played with and cant get to work) and color. Probably better to be using the blue as the secondary, I guess

image
image

### CODE REVIEW

File: packages/common/src/styles/theme.ts [hunk 1]
```diff
@@ -44,6 +44,8 @@ const themeOptions = {
   palette: {
     darkGrey: '#555770',
     primary: { 500: '#e95c30' },
+    secondary: { main: '#555770' },
```
Comment (joshxg) [inline packages/common/src/styles/theme.ts:47]: don't like but can fix with using better svg icons for the checkbox

File: packages/common/src/ui/layout/tables/RemoteDataTable.tsx [hunk 2]
```diff
@@ -28,34 +27,15 @@ import { useTheme } from '@material-ui/core/styles';
 
 import { SortAsc, SortDesc } from '../../icons';
 import { DEFAULT_PAGE_SIZE } from '.';
+import { TableProps } from './types';
+import { useSetupDataTableApi } from './hooks/useDataTableApi';
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/RemoteDataTable.tsx:31]: put the types in a file they were annoying my navigation .. 😁

File: packages/common/src/ui/layout/tables/RemoteDataTable.tsx [hunk 3]
```diff
@@ -67,20 +47,14 @@ export const RemoteDataTable = <T extends Record<string, unknown>>({
   onFetchData,
   onRowClick,
   totalLength = 0,
-}: TableProps<T> & { children?: ReactNode }): JSX.Element => {
+  tableApi,
+}: TableProps<T>): JSX.Element => {
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/RemoteDataTable.tsx:51]: played around with `forwardRef` but couldn't keep the nice generic typing and also made a little more sense to name the prop something other than `ref` (although unsure if `tableApi` is that much better.. )

File: packages/common/src/ui/layout/tables/RemoteDataTable.tsx [hunk 4]
```diff
@@ -137,7 +123,10 @@ export const RemoteDataTable = <T extends Record<string, unknown>>({
               {headers.map(column => (
                 <TableCell
                   {...column.getHeaderProps(column.getSortByToggleProps())}
-                  sx={theme.typography.th}
+                  sx={{
+                    backgroundColor: 'transparent',
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/RemoteDataTable.tsx:127]: the header row was making the drawer shadow go hunchy funchy

File: packages/common/src/ui/layout/tables/RemoteDataTable.tsx [hunk 5]
```diff
@@ -162,7 +151,8 @@ export const RemoteDataTable = <T extends Record<string, unknown>>({
                     <TableCell
                       {...cell.getCellProps()}
                       sx={{
-                        padding: '9.5px',
+                        padding: 0,
+                        paddingLeft: '16px',
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/RemoteDataTable.tsx:155]: I dunno why this started playing funny buggers

File: packages/common/src/ui/layout/tables/columns/CheckboxSelectionColumn.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        color="secondary"
+        size="small"
+        onClick={event => {
+          event.stopPropagation();
+        }}
+        {...row.getToggleRowSelectedProps()}
+      />
+    );
+  },
+});
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/columns/CheckboxSelectionColumn.tsx:26]: this was surprisingly very easy

File: packages/common/src/ui/layout/tables/hooks/useColumns.ts [hunk 7]
```diff
@@ -17,34 +18,45 @@ export interface ColumnDefinition<T> {
...
-  <T extends Record<string, unknown>>(
-    columns: ColumnDefinition<T>[]
-  ): Column<T>[] => {
-    const t = useTranslation();
-    const formatDate = useFormatDate();
+interface ColumnOptions {
+  useSelectionColumn: boolean;
+}
+
+// eslint-disable-next-line @typescript-eslint/ban-types
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/hooks/useColumns.ts:25]: Had some real issues here with types! how annoying!

File: packages/common/src/ui/layout/tables/hooks/useColumns.ts [hunk 8]
```diff
@@ -17,34 +18,45 @@ export interface ColumnDefinition<T> {
...
+export const useColumns = <T extends object>(
+  columnsToMap: ColumnDefinition<T>[],
+  options: ColumnOptions = { useSelectionColumn: false }
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/hooks/useColumns.ts:28]: I don't think this is good, but OK for now, maybe - intention wasn't for a `useX` for every column, but the more generic ones use in different places.. 🤷

File: packages/host/src/Host.tsx [hunk 9]
```diff
@@ -25,7 +25,14 @@ const Content = styled(Box)({
...
   height: '100vh',
 });
-const queryClient = new QueryClient();
+
+const queryClient = new QueryClient({
+  defaultOptions: {
+    queries: {
+      refetchOnWindowFocus: false,
+    },
+  },
```
Comment (joshxg) [inline packages/host/src/Host.tsx:34]: was annoying me that whenever I click dev tools and back on the app it refetches 🤔

File: packages/common/src/ui/layout/tables/columns/CheckboxSelectionColumn.tsx [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { Checkbox } from '../../../components/inputs/Checkbox';
+import { Column } from 'react-table';
+
+export const getCheckboxSelectionColumn = (): Column => ({
+  id: 'selection',
+  Header: ({ getToggleAllRowsSelectedProps }) => (
```
Comment (joshxg) [inline packages/common/src/ui/layout/tables/columns/CheckboxSelectionColumn.tsx:7]: Looks like maybe there isn't a header with a checkbox like this and it'd actually be in some other component in the toolbar above, but left it for now

**Review by mark-prins — APPROVED** [review]
sweeeet!
agree about the ref - not exactly reacty but it works. Not sure about the name though.
The types are a pain, would be nice to revisit those.
I've made some checkbox icons - will do them separately though, so I can add the storybook for icons too.

File: packages/common/src/ui/layout/tables/hooks/useColumns.ts (hunk 8 above)
Comment (joshxg) [reply packages/common/src/ui/layout/tables/hooks/useColumns.ts]: Yup much better
75e26c5

**Pull request merged**